crates/red_knot_test/src/parser.rs

```diff
@@ -381,26 +371,41 @@ impl<'s> Parser<'s> {
             return self.parse_config(code);
         }
 
+        let explicit_path = captures.name("path").map(|it| it.as_str());
+
+        if let Some(explicit_path) = explicit_path {
+            if !lang.is_empty()
+                && explicit_path.contains('.')
+                && !explicit_path.ends_with(&format!(".{lang}"))
+            {
+                return Err(anyhow::anyhow!(
+                    "File `{explicit_path}` has mismatching `lang={lang}`"
+                ));
+            }
+        }
+
+        let path = match explicit_path {
+            Some(path) => path.to_string(),
+            None => {
+                self.unnamed_file_count += 1;
+
+                match lang {
+                    "pyi" => format!("test_{}.pyi", self.unnamed_file_count),
+                    _ => format!("test_{}.py", self.unnamed_file_count),
```

Comment:
Should we match on `"py"` explicitly, and show an error for unsupported extensions?